src/native/http_request_response.c

```diff
@@ -62,50 +62,65 @@ void aws_java_http_stream_from_native_delete(JNIEnv *env, jobject jHttpStream) {
     }
 
-    if (callback->java_http_stream_base) {
-        aws_java_http_stream_from_native_delete(env, callback->java_http_stream_base);
+    if (binding->java_http_response_stream_handler != NULL) {
+        (*env)->DeleteGlobalRef(env, binding->java_http_response_stream_handler);
     }
 
-    if (callback->java_http_response_stream_handler != NULL) {
-        (*env)->DeleteGlobalRef(env, callback->java_http_response_stream_handler);
+    if (binding->native_request) {
+        aws_http_message_release(binding->native_request);
     }
+    aws_byte_buf_clean_up(&binding->headers_buf);
+    aws_mem_release(aws_jni_get_allocator(), binding);
+}
 
-    if (callback->native_request) {
-        aws_http_message_release(callback->native_request);
+void *aws_http_stream_binding_acquire(struct http_stream_binding *binding) {
+    if (binding == NULL) {
+        return NULL;
     }
-    aws_byte_buf_clean_up(&callback->headers_buf);
-    aws_mem_release(aws_jni_get_allocator(), callback);
+    aws_atomic_fetch_add(&binding->ref, 1);
+    return binding;
+}
+
+void *aws_http_stream_binding_release(JNIEnv *env, struct http_stream_binding *binding) {
+    if (binding == NULL) {
+        return NULL;
+    }
+    size_t pre_ref = aws_atomic_fetch_sub(&binding->ref, 1);
+    AWS_ASSERT(pre_ref > 0 && "stream binding refcount has gone negative");
+    if (pre_ref == 1) {
+        s_http_stream_binding_destroy(env, binding);
+    }
+    return NULL;
 }
 
 // If error occurs, A Java exception is thrown and NULL is returned.
-struct http_stream_binding *aws_http_stream_binding_alloc(JNIEnv *env, jobject java_callback_handler) {
+struct http_stream_binding *aws_http_stream_binding_new(JNIEnv *env, jobject java_callback_handler) {
 
     struct aws_allocator *allocator = aws_jni_get_allocator();
-    struct http_stream_binding *callback = aws_mem_calloc(allocator, 1, sizeof(struct http_stream_binding));
-    AWS_FATAL_ASSERT(callback);
+    struct http_stream_binding *binding = aws_mem_calloc(allocator, 1, sizeof(struct http_stream_binding));
+    AWS_FATAL_ASSERT(binding);
 
     // GetJavaVM() reference doesn't need a NewGlobalRef() call since it's global by default
-    jint jvmresult = (*env)->GetJavaVM(env, &callback->jvm);
+    jint jvmresult = (*env)->GetJavaVM(env, &binding->jvm);
     (void)jvmresult;
     AWS_FATAL_ASSERT(jvmresult == 0);
 
-    callback->java_http_response_stream_handler = (*env)->NewGlobalRef(env, java_callback_handler);
-    AWS_FATAL_ASSERT(callback->java_http_response_stream_handler);
-    AWS_FATAL_ASSERT(!aws_byte_buf_init(&callback->headers_buf, allocator, 1024));
+    binding->java_http_response_stream_handler = (*env)->NewGlobalRef(env, java_callback_handler);
+    AWS_FATAL_ASSERT(binding->java_http_response_stream_handler);
+    AWS_FATAL_ASSERT(!aws_byte_buf_init(&binding->headers_buf, allocator, 1024));
 
-    aws_atomic_init_int(&callback->activated, 0);
+    aws_atomic_init_int(&binding->ref, 1);
```

Comment:
just curious, why not use aws_ref_count?

Reply:
The release needs to take `JNIEnv *env`, and I don't think we can save that in the structure?